### PR TITLE
builder: support for ds_bpermute on wave64

### PR DESCRIPTION
**Pull request by OutOfCache**
Add gfx11 to gfxIps supporting ds_bpermute.
Create possibility to use ds_bpermute with wave64
by doing a v_permlane64 and two ds_bpermutes.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
Can one of the admins verify this patch?

**Comment by nhaehnle** [comment]
ok to test

File: llpc/test/shaderdb/general/PipelineVsFs_TestBPermute64.pipe [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+#version 450
+#extension GL_KHR_shader_subgroup_shuffle : enable
+
+void main() {
+    uint index = gl_SubgroupInvocationID;
+    uint value = 4 * index;
+    uint data = subgroupShuffle(value, index ^ 8);
+
+    gl_Position = vec4(0.0, 0.0, data, 1.0);
+}
```
Comment (nhaehnle) [inline llpc/test/shaderdb/general/PipelineVsFs_TestBPermute64.pipe:14]: This test has two weakness.

First, a shuffle by `gl_SubgroupInvocationID ^ 8` should really be implemented entirely differently, using DPP16 or v_permlane16_b32 (maybe one for your todo list?).

Second, there's a lot of noise in the resulting assembly. A better test would avoid that; which is slightly magical, but you can do it with an `.lgc` test like this:
```
define spir_func float @fn(float %value, i32 %index) !lgc.shaderstage !0 {
.entry:
  %r = call float (...) @lgc.create.subgroup.shuffle.f32(float %value, i32 %index)
  ret float %r
}

declare float @lgc.create.subgroup.shuffle.f32(...)

; ShaderStageCompute
!0 = !{i32 7}
```
The RUN line needs to invoke lgc instead of amdllpc.

**Comment by Flakebi** [comment]
For the typo CI, can you add VALU to the `extend-words` list please? https://github.com/GPUOpen-Drivers/llpc/blob/dev/.typos.toml

**Comment by nhaehnle** [comment]
retest this please

**Comment by amdvlk-admin** [comment]
bdc9161192ebfcacaef46c9a75514ae31aff7225 Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/include/llvm-dialects/TableGen/Format.h:77:9: error: ‘Optional’ in namespace ‘llvm’ does not name a template type
   llvm::Optional<llvm::StringRef> getSubstFor(PHKind placeholder) const;
         ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/include/llvm-dialects/TableGen/Format.h:78:9: error: ‘Optional’ in namespace ‘llvm’ does not name a template type
   llvm::Optional<llvm::StringRef> getSubstFor(llvm::StringRef placeholder) const;
         ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:63:1: error: ‘Optional’ does not name a type
 Optional<StringRef>
 ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:74:1: error: ‘Optional’ does not name a type
 Optional<StringRef> FmtContext::getSubstFor(StringRef placeholder) const {
 ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp: In member function ‘void llvm_dialects::FmtObjectBase::format(llvm::raw_ostream&) const’:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:191:9: error: ‘Optional’ was not declared in this scope
         Optional<StringRef> subst;
         ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:191:27: error: expected primary-expression before ‘>’ token
         Optional<StringRef> subst;
                           ^
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:191:29: error: ‘subst’ was not declared in this scope
         Optional<StringRef> subst;
                             ^~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:191:29: note: suggested alternative: ‘strstr’
         Optional<StringRef> subst;
                             ^~~~~
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:194:28: error: ‘const class llvm_dialects::FmtContext’ has no member named ‘getSubstFor’
           subst = context->getSubstFor(repl.spec.substr(1));
                            ^~~~~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:196:28: error: ‘const class llvm_dialects::FmtContext’ has no member named ‘getSubstFor’
           subst = context->getSubstFor(repl.placeholder);
                            ^~~~~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/Format.cpp:211:47: warning: ‘llvm::ArrayRef<T> llvm::makeArrayRef(const std::vector<T>&) [with T = llvm::detail::format_adapter*]’ is deprecated: Use deduction guide instead [-Wdeprecated-declarations]
       auto range = llvm::makeArrayRef(adapters);
                                               ^
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/include/llvm-dialects/TableGen/Format.h:77:9: error: ‘Optional’ in namespace ‘llvm’ does not name a template type
   llvm::Optional<llvm::StringRef> getSubstFor(PHKind placeholder) const;
         ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/include/llvm-dialects/TableGen/Format.h:78:9: error: ‘Optional’ in namespace ‘llvm’ does not name a template type
   llvm::Optional<llvm::StringRef> getSubstFor(llvm::StringRef placeholder) const;
         ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/GenDialect.cpp: In member function ‘std::__cxx11::string {anonymous}::SymbolTable::chooseName(llvm::StringRef)’:
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/imported/llvm-dialects/lib/TableGen/GenDialect.cpp:575:40: warning: ‘llvm::ArrayRef<U> llvm::makeArrayRef(const T&) [with T = llvm::StringRef]’ is deprecated: Use deduction guide instead [-Wdeprecated-declarations]
     return chooseName(makeArrayRef(name));

**Comment by OutOfCache** [comment]
> For the typo CI, can you add VALU to the `extend-words` list please? https://github.com/GPUOpen-Drivers/llpc/blob/dev/.typos.toml

Sure! Should this happen in a separate PR on in this one?

**Comment by OutOfCache** [comment]
I see, I will do that then

File: llpc/test/shaderdb/general/PipelineVsFs_TestBPermute64.pipe (hunk 1 above)
Comment (OutOfCache) [reply llpc/test/shaderdb/general/PipelineVsFs_TestBPermute64.pipe:14]: Is there a way to set the Subgroup size to 64 in this case?

The compute shader usually sets the waveSize to 64 (apparently there were some tests that confirmed a performance increase in that case). 
There are two ways to overwrite this: set the option `useSubgroupSize`. As far as I can tell, this happens in the `SPIRVReader.cpp`. I am not sure if this is called when calling lgc directly, though.
And the other option is by making the `workgroupSize` <= 32. In that case, the waveSize is also reduced to 32. 

Is there a way to make the workgroup > 32 here?

**Comment by amdvlk-admin** [comment]
### Test summary for commit 3825660afd51ba2a25afb5d6f04dd593d3fde677
<details><summary>CTS tests (Failed: 119/211487)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39129/70495 (55.5%)</li>
<li><details><summary>Failed: 119/70495 (0.2%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_double_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_f16vec4_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_float16_t_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_i8vec3_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_int16_t
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_int8_t
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_u16vec3_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_u64vec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffledown_f16vec3
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffledown_i16vec4_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffledown_i64vec2_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffledown_i64vec4_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffledown_ivec3_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffledown_u64vec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffledown_u8vec2_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffledown_uvec3
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_bvec3
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_bvec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_double
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_dvec3_requiredsubgroupsize
Stack trace: Fail
...
```
</p></details></li>

<li>Not Supported: 31247/70495 (44.3%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39498/70495 (56.0%)</li>
<li>Failed: 0/70495 (0.0%)</li>
<li>Not Supported: 30997/70495 (44.0%)</li><li>Warnings: 0/70495 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36345/70497 (51.6%)</li>
<li>Failed: 0/70497 (0.0%)</li>
<li>Not Supported: 34152/70497 (48.4%)</li><li>Warnings: 0/70497 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — COMMENTED** [review]
Thanks, the code looks right to me but it looks like you need to update the test.

By the way, you should remove the --check-pal-metadata. I don't think it adds anything useful to the test in this case.

**Comment by amdvlk-admin** [comment]
### Test summary for commit 9aef301eb8e278ab16c2d014bd77e3d7e9ecfade
<details><summary>CTS tests (Failed: 67/211523)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 39181/70507 (55.6%)</li>
<li><details><summary>Failed: 67/70507 (0.1%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_double_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_f16vec4_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_float16_t_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_i8vec3_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_int16_t
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_int8_t
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_u16vec3_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_u64vec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_bvec3
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_bvec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_double
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_dvec3_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_i64vec4_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_ivec2
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_ivec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_u16vec2
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_uint16_t
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshufflexor_bvec2_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshufflexor_dvec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshufflexor_float
Stack trace: Fail
...
```
</p></details></li>

<li>Not Supported: 31259/70507 (44.3%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 39498/70507 (56.0%)</li>
<li>Failed: 0/70507 (0.0%)</li>
<li>Not Supported: 31009/70507 (44.0%)</li><li>Warnings: 0/70507 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 36345/70509 (51.5%)</li>
<li>Failed: 0/70509 (0.0%)</li>
<li>Not Supported: 34164/70509 (48.5%)</li><li>Warnings: 0/70509 (0.0%)</li>
</ul></details>


</ul></details>

File: llpc/test/shaderdb/relocatable_shaders/DescPtrSingleSelect.spvasm [hunk 2]
```diff
@@ -2,7 +2,10 @@
 
 ; BEGIN_SHADERTEST
 ; RUN: amdllpc -o %t.elf -gfxip=9 -amdgpu-transform-discard-to-demote %s && llvm-objdump --triple=amdgcn --mcpu=gfx900 -d %t.elf | FileCheck -check-prefix=SHADERTEST %s
+<<<<<<< HEAD
 ; REQUIRES: do-not-run-me
+=======
+>>>>>>> f17f0871 (Update llpc from commit 978d8327)
```
Comment (Flakebi) [inline llpc/test/shaderdb/relocatable_shaders/DescPtrSingleSelect.spvasm:8]: There’s a merge conflict in this file

**Comment by amdvlk-admin** [comment]
### Test summary for commit 0fe550f43510c9b1c1b625e31eced9501ca62432
<details><summary>CTS tests (Failed: 55/205070)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36286/68141 (53.3%)</li>
<li><details><summary>Failed: 55/68141 (0.1%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_bvec3
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_dvec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_i16vec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_ivec4_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffle_uvec3_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_bvec2
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_dvec2_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_dvec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_dvec4_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_i16vec3_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_i64vec2
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_i64vec2_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_i8vec2_requiredsubgroupsize
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_u64vec3
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_uint
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_uint64_t
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshuffleup_uvec2
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshufflexor_bool
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshufflexor_bvec4
Stack trace: Fail

FAILURE: dEQP-VK.subgroups.shuffle.compute.subgroupshufflexor_dvec2_requiredsubgroupsize
Stack trace: Fail
...
```
</p></details></li>

<li>Not Supported: 31800/68141 (46.7%)</li><li>Warnings: 0/68141 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 36569/68141 (53.7%)</li>
<li>Failed: 0/68141 (0.0%)</li>
<li>Not Supported: 31572/68141 (46.3%)</li><li>Warnings: 0/68141 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34225/68788 (49.8%)</li>
<li>Failed: 0/68788 (0.0%)</li>
<li>Not Supported: 34563/68788 (50.2%)</li><li>Warnings: 0/68788 (0.0%)</li>
</ul></details>


</ul></details>

File: include/vkgcDefs.h [hunk 3]
```diff
@@ -44,7 +44,7 @@
 #endif
 
 /// LLPC major interface version.
-#define LLPC_INTERFACE_MAJOR_VERSION 60
+#define LLPC_INTERFACE_MAJOR_VERSION 58
```
Comment (nhaehnle) [inline include/vkgcDefs.h:47]: This change looks wrong / unrelated. Maybe a rebase artefact?

File: llpc/test/shaderdb/general/PipelineCs_StrideReloc.pipe [hunk 4]
```diff
@@ -2,7 +2,7 @@
 ; BEGIN_SHADERTEST
 ; RUN: amdllpc -enable-relocatable-shader-elf -o %t.elf %gfxip %s && llvm-objdump --triple=amdgcn --mcpu=gfx900 -d %t.elf | FileCheck -check-prefix=SHADERTEST %s
 ; REQUIRES: do-not-run-me
-  // Test that correct stride is in the linked texture fetching code.
+  // Test that correct relocated offsets are in the linked texture fetching code.
```
Comment (nhaehnle) [inline llpc/test/shaderdb/general/PipelineCs_StrideReloc.pipe:5]: Another unrelated change.

**Comment by tsymalla** [comment]
Can you also please fix the typo in the commit message? ds_bperumute => ds_bpermute

**Comment by amdvlk-admin** [comment]
### Test summary for commit d19c35bcc36c8e8fa44606d60d5f2ccefc2f131d
<details><summary>CTS tests (Failed: 0/207172)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36989/69064 (53.6%)</li>
<li>Failed: 0/69064 (0.0%)</li>
<li>Not Supported: 32075/69064 (46.4%)</li><li>Warnings: 0/69064 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37209/69042 (53.9%)</li>
<li>Failed: 0/69042 (0.0%)</li>
<li>Not Supported: 31833/69042 (46.1%)</li><li>Warnings: 0/69042 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34228/69066 (49.6%)</li>
<li>Failed: 0/69066 (0.0%)</li>
<li>Not Supported: 34838/69066 (50.4%)</li><li>Warnings: 0/69066 (0.0%)</li>
</ul></details>


</ul></details>

**Comment by nhaehnle** [comment]
In the future, please don't merge dev into your branch; rebase instead if necessary.

File: lgc/builder/SubgroupBuilder.cpp [hunk 5]
```diff
@@ -325,6 +325,38 @@ Value *SubgroupBuilder::CreateSubgroupShuffle(Value *const value, Value *const i
     // The ds_bpermute intrinsic requires the index be multiplied by 4.
     return CreateMapToInt32(mapFunc, value, CreateMul(index, getInt32(4)));
   }
+
+  if (supportPermLane64Dpp()) {
+    assert(getShaderWaveSize() == 64);
+
+    // Start the WWM section by setting the inactive lanes.
+    Value *const identity = Constant::getNullValue(value->getType());
+    Value *wwmValue = BuilderBase::get(*this).CreateSetInactive(value, identity);
```
Comment (nhaehnle) [inline lgc/builder/SubgroupBuilder.cpp:334]: What happens if you use PoisonValue instead of a null value? Reading an inactive lane is undefined, so semantically it ought to be okay.

**Review by nhaehnle — COMMENTED** [review]
Conceptually, shouldn't `index` also go through `amdgcn.set.inactive`?

The other part of it: Via the scratch instructions, you can see that there are currently 4 registers that are treated as WWM registers. It would be good to reduce that.

I think one way to do that is to apply the `amdgcn.wwm` directly to the bpermute results, since, the final select doesn't need to run in WWM, pull the index multiply by 4 to above the `amdgcn.set.inactive`, and explicitly do `amdgcn.set.inactive` on both the value and the scaled index, both with a `poison` inactive value.

**Comment by amdvlk-admin** [comment]
1026ca560cd6cdda470d555b2b47b43af9c1bf55 Jenkins build error.
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchBufferOp.cpp:585:21: error: redeclaration of ‘llvm::Constant* const nullPointer’
     Constant *const nullPointer = ConstantPointerNull::get(m_offsetType);
                     ^~~~~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchBufferOp.cpp:584:21: note: ‘llvm::Constant* const nullPointer’ previously declared here
     Constant *const nullPointer = ConstantPointerNull::get(m_offsetType);
                     ^~~~~~~~~~~
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchBufferOp.cpp:771:18: error: redeclaration of ‘llvm::PHINode* const newPhiNode’
   PHINode *const newPhiNode = m_builder->CreatePHI(m_offsetType, incomings.size());
                  ^~~~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchBufferOp.cpp:770:18: note: ‘llvm::PHINode* const newPhiNode’ previously declared here
   PHINode *const newPhiNode = m_builder->CreatePHI(m_offsetType, incomings.size());
                  ^~~~~~~~~~
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchBufferOp.cpp:1200:13: error: redeclaration of ‘llvm::Constant* nullPointer’
   Constant *nullPointer = ConstantPointerNull::get(m_offsetType);
             ^~~~~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchBufferOp.cpp:1199:13: note: ‘llvm::Constant* nullPointer’ previously declared here
   Constant *nullPointer = ConstantPointerNull::get(m_offsetType);
             ^~~~~~~~~~~
--
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchInOutImportExport.cpp:3490:17: error: redeclaration of ‘llvm::CallInst* readCall’
       CallInst *readCall = cast<CallInst>(output);
                 ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchInOutImportExport.cpp:3487:17: note: ‘llvm::CallInst* readCall’ previously declared here
       CallInst *readCall = cast<CallInst>(output);
                 ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchInOutImportExport.cpp:3492:16: error: redeclaration of ‘unsigned int location’
       unsigned location = cast<ConstantInt>(readCall->getArgOperand(0))->getZExtValue();
                ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchInOutImportExport.cpp:3489:16: note: ‘unsigned int location’ previously declared here
       unsigned location = cast<ConstantInt>(readCall->getArgOperand(0))->getZExtValue();
                ^~~~~~~~
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchInOutImportExport.cpp:3501:106: error: expected ‘)’ before ‘;’ token
                         {Attribute::Speculatable, Attribute::ReadOnly, Attribute::WillReturn}, insertPos);
                                                                                                          ^
/jenkins/workspace/vulkan/sanitized-opensource/Github-PR/llpc-github-pr/driver_build/drivers/llpc/lgc/patch/PatchInOutImportExport.cpp:3508:106: error: expected ‘)’ before ‘;’ token
                         {Attribute::Speculatable, Attribute::ReadOnly, Attribute::WillReturn}, insertPos);
                                                                                                          ^
[2883/3038] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/patch/PatchPeepholeOpt.cpp.o
[2884/3038] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/patch/PatchLoadScalarizer.cpp.o
[2885/3038] Building CXX object compiler/llpc/llvm/tools/lgc/CMakeFiles/LLVMlgc.dir/patch/NggPrimShader.cpp.o

**Comment by amdvlk-admin** [comment]
### Test summary for commit 377db76963c7bdfd2241ec54d349dc37f7283971
<details><summary>CTS tests (Failed: 0/207172)</summary><ul><li>Built with version 1.3.0.0</li>
<details open><summary>Ubuntu 22.04, Navi3x</summary><ul>

<li>Passed: 36989/69064 (53.6%)</li>
<li>Failed: 0/69064 (0.0%)</li>
<li>Not Supported: 32075/69064 (46.4%)</li><li>Warnings: 0/69064 (0.0%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Navi2x</summary><ul>

<li>Passed: 37209/69042 (53.9%)</li>
<li>Failed: 0/69042 (0.0%)</li>
<li>Not Supported: 31833/69042 (46.1%)</li><li>Warnings: 0/69042 (0.0%)</li>
</ul></details>

<details open><summary>Rhel 9.0, Gfx10</summary><ul>

<li>Passed: 34228/69066 (49.6%)</li>
<li>Failed: 0/69066 (0.0%)</li>
<li>Not Supported: 34838/69066 (50.4%)</li><li>Warnings: 0/69066 (0.0%)</li>
</ul></details>


</ul></details>

**Review by nhaehnle — APPROVED** [review]
Thanks, the generated ISA definitely looks better.
